### PR TITLE
docs: add Zeabur as a hosting provider

### PR DESCRIPTION
**Pull request by MichaelYuhe**
[Zeabur](https://zeabur.com) is a platform for hosting full stack apps, including Telegram bots.
This guide introduces how to deploy grammY bots through Zeabur's web dashboard or CLI.

### CODE REVIEW

File: site/docs/hosting/zeabur.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+# Hosting: Zeabur
+
+This tutorial will guide you how to deploy your grammY bots to [Zeabur](https://zeabur.com/).
+
+## Prerequisites
+
+To follow along, you need to have a [Github](https://github.com) and [Zeabur](https://zeabur.com/) account.
+
+### Method1: Create a New Project from Scratch
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur.md:14]:
```suggestion
### Method 1: Create a New Project from Scratch
```

File: site/docs/hosting/zeabur.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+### Method 2: Use Zeabur's Template
+
+Zeabur has already provided a template for you to use.
+You can find it [here](https://github.com/zeabur/telegram-bot-starter).
+
+You can just use the template and start writing your bot's code.
+
+## Deploying
+
+### Method 1: Deploy from GitHub in Zeabur's Dashboard
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur.md:104]:
```suggestion
### Method 1: Deploy From GitHub in Zeabur's Dashboard
```

File: site/docs/hosting/zeabur.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+1. Create a repository on GitHub, it can be public or private and push your code to it.
+2. Go to [Zeabur dashboard](https://dash.zeabur.com).
+3. Click on the `New Project` button, and click on the `Deploy New Service` button, choose `GitHub` as the source and select your repository.
+4. Go to Variables tab to add your environment variables like `BOT_TOKEN`.
+5. Your service will be deployed automatically.
+
+### Method 2: Deploy with Zeabur CLI
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur.md:112]:
```suggestion
### Method 2: Deploy With Zeabur CLI
```

File: site/docs/hosting/zeabur.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+# Hosting: Zeabur
+
+This tutorial will guide you how to deploy your grammY bots to [Zeabur](https://zeabur.com/).
+
+## Prerequisites
+
+To follow along, you need to have a [Github](https://github.com) and [Zeabur](https://zeabur.com/) account.
+
+### Method 1: Create a New Project from Scratch
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur.md:14]:
```suggestion
### Method 1: Create a New Project From Scratch
```

File: site/docs/hosting/zeabur.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+1. Create a repository on GitHub, it can be public or private and push your code to it.
+2. Go to [Zeabur dashboard](https://dash.zeabur.com).
+3. Click on the `New Project` button, and click on the `Deploy New Service` button, choose `GitHub` as the source and select your repository.
+4. Go to Variables tab to add your environment variables like `BOT_TOKEN`.
+5. Your service will be deployed automatically.
+
+### Method 2: Deploy With Zeabur CLI
+
+cd into your project directory and run the following command:
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur.md:114]: Does this look better?
```suggestion
`cd` into your project directory and run the following command:
```

File: site/docs/hosting/zeabur.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+```sh
+npx @zeabur/cli deploy
+```
+
+Follow the instructions to select a region to deploy, and your bot will be deployed automatically.
+
+### Setting the Webhook URL
+
+If you are using webhooks, after your first deployment, you should configure your bot's webhook settings to point to your app.
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur.md:124]: This implies that it's also possible not to use webhooks. Aren't you building some sort of serverless platform that responds to web requests only? (I'd be curious how you would manage to do polling.)

**Review by KnorpelSenf — CHANGES_REQUESTED** [review]
You install grammY and express, but people would first have to go to the webhooks guide and see what's necessary to build a bot. It would be cool to have a code snippet with a minimal echo bot in this guide. That way, people don't have to leave the page and can just copy and paste everything.

**Comment by MichaelYuhe** [comment]
> You install grammY and express, but people would first have to go to the webhooks guide and see what's necessary to build a bot. It would be cool to have a code snippet with a minimal echo bot in this guide. That way, people don't have to leave the page and can just copy and paste everything.

I take reference from other guides, maybe it's better to provide my own code snippets in [the template on Zeabur](https://github.com/zeabur/telegram-bot-starter)?  
image

**Comment by KnorpelSenf** [comment]
I have seen that Zeabur can host Deno code, too: https://zeabur.com/docs/guides/deno/typescript
Should we directly add two different guides, much like what we have for CFW?

**Comment by KnorpelSenf** [comment]
> > You install grammY and express, but people would first have to go to the webhooks guide and see what's necessary to build a bot. It would be cool to have a code snippet with a minimal echo bot in this guide. That way, people don't have to leave the page and can just copy and paste everything.
> 
> I take reference from other guides

Which other guide is missing a code snippet?

Either way, just because other parts of the docs can be improved, does not mean that we can't do it better here :)

**Comment by MichaelYuhe** [comment]
> I have seen that Zeabur can host Deno code, too: https://zeabur.com/docs/guides/deno/typescript Should we directly add two different guides, much like what we have for CFW?
Yeah that sounds good, I can add the other guide later

**Comment by MichaelYuhe** [comment]
BTW, why do you prefer using `From` instead of `from` in the title?

**Comment by MichaelYuhe** [comment]
> > > You install grammY and express, but people would first have to go to the webhooks guide and see what's necessary to build a bot. It would be cool to have a code snippet with a minimal echo bot in this guide. That way, people don't have to leave the page and can just copy and paste everything.
> > 
> > 
> > I take reference from other guides
> 
> Which other guide is missing a code snippet?
> 
> Either way, just because other parts of the docs can be improved, does not mean that we can't do it better here :)

New lesson learned today :D

**Comment by MichaelYuhe** [comment]
I just added a deno guide and simplify the code demo, could u please review agin? Thanks

File: site/docs/hosting/zeabur-deno.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+This tutorial will guide you how to deploy your grammY bots with Deno to [Zeabur](https://zeabur.com/).
+
+::: tip Looking for the Node.js Version?
+This tutorial explains how to deploy a Telegram bot to Zeabur using Deno.
+If you're looking for the Node.js version, please check out [this tutorial](./zeabur-nodejs) instead.
+:::
+
+## Prerequisites
+
+To follow along, you need to have a [Github](https://github.com) and [Zeabur](https://zeabur.com/) account.
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-deno.md:19]:
```suggestion
To follow along, you need to have [Github](https://github.com) and [Zeabur](https://zeabur.com/) accounts.
```

File: site/docs/hosting/zeabur-nodejs.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+::: tip Looking for the Deno Version?
+This tutorial explains how to deploy a Telegram bot to Zeabur using Node.js.
+If you're looking for the Deno version, please check out [this tutorial](./zeabur-deno) instead.
+:::
+
+
+## Prerequisites
+
+To follow along, you need to have a [Github](https://github.com) and [Zeabur](https://zeabur.com/) account.
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-nodejs.md:20]:
```suggestion
To follow along, you need to have [Github](https://github.com) and [Zeabur](https://zeabur.com/) accounts.
```

File: site/docs/hosting/zeabur-nodejs.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+pnpm init -y
+
+# Install main dependencies.
+pnpm install grammy
+
+# Install development dependencies.
+pnpm install -D nodemon
+```
+
+Then, cd into `src/`, and create a file named `bot.js`.
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-nodejs.md:39]: I actually liked it that you directly illustrated a TypeScript setup. Most people will want to do this anyway and the others (really should switch but if they don't want to then they) can just skip reading that part. Having a slightly longer page is better than leaving it to the reader to figure out how to do the setup themselves, what do you think?

File: site/docs/hosting/zeabur-deno.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+# Create deno.json file to generate lock file
+touch deno.json
+```
+
+Then modify `main.ts` file with the following code:
+
+```typescript
+import { Bot } from "https://deno.land/x/grammy@v1.23.0/mod.ts";
+
+const bot = new Bot("your_telegram_bot_token");
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-deno.md:44]: The Node version reads the token from the environment. Is there a reason why this isn't done here?

File: site/docs/hosting/zeabur-nodejs.md [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+  
+  const response = "Hello, I'm a bot!";
+
+  ctx.reply(response);
+});
+
+bot.start();
+```
+
+> Note: Replace `your_telegram_bot_token` with your bot's token from the BotFather.
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-nodejs.md:62]: This isn't correct, is it? It should tell people to set an env var.

**Review by KnorpelSenf — CHANGES_REQUESTED** [review]
Can you make sure to only write one sentence per line? We have found this to be a great rule that helps us preserve readable git diffs.

**Comment by MichaelYuhe** [comment]
Just updated the variable docs and make it more readable

**Comment by KnorpelSenf** [comment]
Looking pretty good to me! What are we going to do with the TypeScript part? Would you like to add instructions on a TS setup again, or do you have a reason why we should leave a TS setup to users?

Also, there are a few minor nitpicks:
- please fix the formatting by running `cd site && npm run docs:fmt`
- please fix the linting
- please fix the following two typos:

File: site/docs/hosting/zeabur-deno.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+const bot = new Bot(token); 
+
+bot.command("start", (ctx) => ctx.reply("Hello from Deno & grammY!"));
+
+bot.on("message", (ctx) => ctx.reply("How can I help you?"));
+
+bot.start();
+```
+
+> Note: Get your bot token with [BotFather on telegram](https://t.me/BotFather), and set is as an environment variable `TELEGRAM_BOT_TELEGRAM` in Zeabur.
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-deno.md:56]:
```suggestion
> Note: Get your bot token with [@BotFather](https://t.me/BotFather) on Telegram, and set is as an environment variable `TELEGRAM_BOT_TELEGRAM` in Zeabur. 
```
```suggestion
> Note: Get your bot token with [BotFather on telegram](https://t.me/BotFather), and set is as an environment variable `TELEGRAM_BOT_TELEGRAM` in Zeabur. 
```

File: site/docs/hosting/zeabur-nodejs.md (hunk 9 above)
Comment (MichaelYuhe) [reply site/docs/hosting/zeabur-nodejs.md:39]: I also think using ts will be better for this project, I need to modify the template on Zeabur first and then updates the docs. Thanks for advice

File: site/docs/.vitepress/configs/locales/en.ts [hunk 13]
```diff
@@ -392,6 +392,18 @@ const hostingTutorials = {
...
+      link: "/hosting/zeabur-deno",
+    },
+    {
+      text: "Zeabur (Node.js)",
+      link: "/hosting/zeabur-nodejs",
+    },
+    {
+      text: "Virtual Private Server",
+      link: "/hosting/vps",
+    }
```
Comment (roj1512) [inline site/docs/.vitepress/configs/locales/en.ts:406]: What is this addition? Merge error?

**Comment by roj1512** [comment]
Waiting on https://github.com/grammyjs/website/pull/1059#issuecomment-2129073823

**Comment by roj1512** [comment]
@MichaelYuhe Any updates?

**Comment by MichaelYuhe** [comment]
> @MichaelYuhe Any updates?

Just updated the template to use typescript, now I will update and format the docs

File: site/docs/hosting/zeabur-deno.md [hunk 14]
```diff
@@ -0,0 +1,92 @@
+---
+prev: false
+next: false
+---
+
+# Hosting: Zeabur (Deno)
+
+[Zeabur](https://zeabur.com) is a platform that allows you to deploy your full-stack applications with ease. It supports various programming languages and frameworks, including Deno and grammY.
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-deno.md:8]: Please stick with one sentence per line
```suggestion
[Zeabur](https://zeabur.com) is a platform that allows you to deploy your full-stack applications with ease.
It supports various programming languages and frameworks, including Deno and grammY.
```

File: site/docs/hosting/zeabur-deno.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+touch main.ts
+
+# Create deno.json file to generate lock file
+touch deno.json
+```
+
+Then modify `main.ts` file with the following code:
+
+```typescript
+import { Bot } from "https://deno.land/x/grammy@v1.23.0/mod.ts";
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-deno.md:42]: The latest version number will actually be inserted during build. Just use a bare link instead.
```suggestion
import { Bot } from "https://deno.land/x/grammy/mod.ts";
```

File: site/docs/hosting/zeabur-deno.md [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+>
+> You can check out [this tutorial](https://zeabur.com/docs/deploy/variables) for setting environment variables in Zeabur.
+
+Then run the following command to start your bot:
+
+```sh
+deno run --allow-net main.ts
+```
+
+Deno will automatically download the dependencies, generate the lock file and start your bot.
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-deno.md:66]: Style nitpick (both works but I like this better)
```suggestion
Deno will automatically download the dependencies, generate the lock file, and start your bot.
```

File: site/docs/hosting/zeabur-nodejs.md [hunk 17]
```diff
@@ -0,0 +1,132 @@
+---
+prev: false
+next: false
+---
+
+# Hosting: Zeabur (Node.js)
+
+[Zeabur](https://zeabur.com) is a platform that allows you to deploy your full-stack applications with ease. It supports various programming languages and frameworks, including Node.js and grammY.
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-nodejs.md:8]: One sentence per line
```suggestion
[Zeabur](https://zeabur.com) is a platform that allows you to deploy your full-stack applications with ease.
It supports various programming languages and frameworks, including Node.js and grammY.
```

File: site/docs/hosting/zeabur-nodejs.md [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+# Initialize TypeScript.
+npx tsc --init
+```
+
+Then, cd into `src/`, and create a file named `bot.ts`.
+It is where you will write your bot's code.
+
+Now, you can start writing your bot's code in `src/bot.ts`.
+
+```js
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-nodejs.md:46]:
```suggestion
```ts
```

File: site/docs/hosting/zeabur-nodejs.md [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+const bot = new Bot(
+  process.env.TELEGRAM_BOT_TOKEN || "your_telegram_bot_token",
+);
+
+bot.on("message:text", async (ctx) => {
+  console.log("Message: ", ctx.message.text);
+
+  const response = "Hello, I'm a bot!";
+
+  ctx.reply(response);
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-nodejs.md:58]:
```suggestion
  await ctx.reply(response);
```

File: site/docs/hosting/zeabur-nodejs.md [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+```
+
+And then we have to add `start` scripts to our `package.json`.
+Our `package.json` should now be similar to this:
+
+```json
+{
+  "name": "telegram-bot-starter",
+  "version": "1.0.0",
+  "description": "Telegram Bot Starter with JavaScript and Grammy",
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-nodejs.md:86]:
```suggestion
  "description": "Telegram Bot Starter with TypeScript and grammY",
```

File: site/docs/hosting/zeabur-nodejs.md [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+  "scripts": {
+    "start": "ts-node src/bot.ts"
```
Comment (KnorpelSenf) [inline site/docs/hosting/zeabur-nodejs.md:88]: This cannot work if `ts-node` is not installed?

**Review by KnorpelSenf — CHANGES_REQUESTED** [review]
Just a few minor things. This looks very good to me overall!

File: site/docs/hosting/zeabur-nodejs.md (hunk 21 above)
Comment (MichaelYuhe) [reply site/docs/hosting/zeabur-nodejs.md:89]: oh my bad for not considering this situation, gonna fix it soon

Comment (MichaelYuhe) [reply site/docs/hosting/zeabur-nodejs.md:89]: Just updated the doc, can u help check again? :D Thanks for the patient review!

**Review by KnorpelSenf — APPROVED** [review]
Awesome stuff! Thank you so much!

**Review by roj1512 — APPROVED** [review]

**Comment by KnorpelSenf** [comment]
@MichaelYuhe thanks! You're done here now. We just need to wait for the translation team to sync the changes across languages and then we'll merge.

**Comment by niusia-ua** [comment]
@KnorpelSenf Can you merge this into a temporary branch so we can do a translation?

**Comment by KnorpelSenf** [comment]
@niusia-ua at this point I'd like to automate it.

**Comment by KnorpelSenf** [comment]
It turns out that spending just 2 hours on an automation is not enough, and doing it manually only takes 30 seconds. grammY must exist for many more decades before such an automation would be worth it.

**Comment by niusia-ua** [comment]
Haha, automation is a typical problem for programmers 😁

**Pull request closed without merging**